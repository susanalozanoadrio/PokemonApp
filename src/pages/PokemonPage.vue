<template>
<div class="container">
  <h1 v-if="!pokemon">Espere por favor ...</h1>
  <div v-else>
    <h1 class="title">¿Quién es este pokémon?</h1>
    <!-- TODO: img -->
    <!-- TODO: Opciones -->
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon"/>
    <PokemonOptions 
      :pokemons="pokemonArr"
      @selection="checkAnswer"
      />
    <template v-if="showAnswer">
      <div  class="container-button">
        <h2 class="fade-in message">{{message}}</h2>
        <button class="myBtn" @click="newGame">Nuevo Juego</button>
      </div>
    </template>
  </div>
  </div>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture'
import PokemonOptions from '@/components/PokemonOptions'
import getPokemonOptions from '@/helpers/getPokemonOptions'

//console.log(getPokemonOptions())

export default {
   components: {
    PokemonPicture,
    PokemonOptions,
  },
  data(){
    return{
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: ''
    }
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonOptions()
      const rndInt = Math.floor( Math.random() * 4)
      this.pokemon = this.pokemonArr[ rndInt ]
      console.log(this.pokemonArr)
      console.log(rndInt)
    },
    checkAnswer(pokemonId ) {
      this.showPokemon = true
      this.showAnswer = true
      console.log('pokemon page llamado', pokemonId)
      if(pokemonId === this.pokemon.id) {
        this.message = `Correcto, es ${this.pokemon.name} :)`
      }else {
        this.message = `Oops, era ${this.pokemon.name} :(`
      }
    },
    newGame(){
      this.showPokemon = false
      this.showAnswer = false
      this.pokemonArr = []
      this.pokemon = null
      this.mixPokemonArray()
    }
  },
  mounted() {
    this.mixPokemonArray()
  }
}

</script>

<style>

.container {
  width: 60%;
  padding: 0% 0px;
  border: 1px solid black;
  background-color: #f1f1f1;
}

.title {
  margin-bottom: 8%;
  color: rebeccapurple;
}

.message {
  color: rebeccapurple;
}

.container-button {
  width: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.myBtn {
  width: 200px;
  padding: 2%;
  color: white;
  font-family: 'Suez One', serif;
  background-color: rebeccapurple;
  border: none;
  border-radius: 4px;
  margin-bottom: 5%;
  
}

</style>