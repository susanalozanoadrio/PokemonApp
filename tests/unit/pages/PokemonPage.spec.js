import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import { mockPokemons } from '../mocks/pokemons.mock'

describe('PokemonPage pages', () => {

    let wrapper
    beforeEach(() => {
        wrapper = shallowMount( PokemonPage )
    })

    test('debe hacer match con el snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })
    test('debe de llamar mixPokemonArray al montar', () => {

        const mixPokemonArraySpy = jest.spyOn( PokemonPage.methods, 'mixPokemonArray')
        const wrapper = shallowMount( PokemonPage )
        expect( mixPokemonArraySpy ).toHaveBeenCalled
    })
    test('debe hacer match con el snapshot cuando cargan los pokemon', () => {

        const wrapper = mount( PokemonPage, {
            data(){
                return{
                  pokemonArr: mockPokemons,
                  pokemon: mockPokemons[0],
                  showPokemon: false,
                  showAnswer: false,
                  message: ''
                }
              }
        })

        expect( wrapper.html() ).toMatchSnapshot()
    })
    test('debe mostrar los componentes PokemonPicture y PokemonOption', () => {

        const wrapper = shallowMount( PokemonPage, {
            data(){
                return{
                  pokemonArr: mockPokemons,
                  pokemon: mockPokemons[0],
                  showPokemon: false,
                  showAnswer: false,
                  message: ''
                }
              }
        })

        //console.log(wrapper.html())
        const picture = wrapper.find('pokemon-picture-stub')
        const options = wrapper.find('pokemon-options-stub')
        expect(picture.exists()).toBeTruthy()
        expect(options.exists()).toBeTruthy()
        
        expect(picture.attributes('pokemonid')).toEqual('1')
        
        expect(options.attributes('pokemons')).toBeTruthy()
    })
    test('pruebas con checkAnswer', async() => {

        const wrapper = shallowMount( PokemonPage, {
            data(){
                return{
                  pokemonArr: mockPokemons,
                  pokemon: mockPokemons[0],
                  showPokemon: false,
                  showAnswer: false,
                  message: ''
                }
              }
        })

        await wrapper.vm.checkAnswer(1)
        expect(wrapper.find('h2').exists()).toBeTruthy()
        //console.log(wrapper.find('h2').text())
        expect(wrapper.vm.showPokemon).toBeTruthy()
        expect(wrapper.find('h2').text()).toBe(`Correcto, es ${mockPokemons[0].name} :)`)

        await wrapper.vm.checkAnswer(4)
        //console.log(wrapper.find('h2').text())
        expect(wrapper.vm.message).toBe(`Oops, era ${mockPokemons[0].name} :(`)


    })
})