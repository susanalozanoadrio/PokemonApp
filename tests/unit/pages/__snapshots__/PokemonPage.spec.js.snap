// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage pages debe hacer match con el snapshot 1`] = `
<div class="container">
  <h1>Espere por favor ...</h1>
</div>
`;

exports[`PokemonPage pages debe hacer match con el snapshot cuando cargan los pokemon 1`] = `
<div class="container">
  <div>
    <h1 class="title">¿Quién es este pokémon?</h1><!-- TODO: img -->
    <!-- TODO: Opciones -->
    <div class="pokemon-container"><img src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg" alt="pokemon" class="hidden-pokemon"></div>
    <div class="options-container">
      <ul>
        <li>bulbasaur</li>
        <li>ivysaur</li>
        <li>venusaur</li>
        <li>charmander</li>
      </ul>
    </div>
    <!--v-if-->
  </div>
</div>
`;
